cmd/ocsp-updater/main.go

```diff
@@ -51,28 +102,70 @@ func setupClients(c cmd.Config, stats statsd.Statter) (rpc.CertificateAuthorityC
 	return cac, closeChan
 }
 
-func (updater *OCSPUpdater) processResponse(tx *gorp.Transaction, serial string) error {
-	certObj, err := tx.Get(core.Certificate{}, serial)
-	if err != nil {
-		return err
-	}
-	statusObj, err := tx.Get(core.CertificateStatus{}, serial)
-	if err != nil {
-		return err
+var errNoStaleResponses = errors.New("No stale responses to update")
+
+func (updater *OCSPUpdater) findStaleOCSPResponses(oldestLastUpdatedTime time.Time, batchSize int) ([]core.Certificate, error) {
+	var certs []core.Certificate
+	_, err := updater.dbMap.Select(
+		&certs,
+		`SELECT cert.*
+		 FROM certificateStatus AS cs
+		 JOIN certificates AS cert
+		 ON cs.serial = cert.serial
+		 WHERE cs.ocspLastUpdated < :lastUpdate
+		 AND cert.expires > now()
+		 ORDER BY cs.ocspLastUpdated ASC
+		 LIMIT :limit`,
+		map[string]interface{}{
+			"lastUpdate": oldestLastUpdatedTime,
+			"limit":      batchSize,
+		},
+	)
+	if err == sql.ErrNoRows || len(certs) == 0 {
+		return certs, errNoStaleResponses
 	}
+	return certs, err
+}
 
-	cert, ok := certObj.(*core.Certificate)
-	if !ok {
-		return fmt.Errorf("Cast failure")
+var errNoNewCertificates = errors.New("No certificates issued since last check")
+
+func (updater *OCSPUpdater) getCertificatesIssuedSince(since time.Time, batchSize int) ([]core.Certificate, error) {
+	var certs []core.Certificate
+	_, err := updater.dbMap.Select(
+		&certs,
+		`SELECT * FROM certificates
+		 WHERE issued > :since
```

Comment:
Also, I would like this to take a pair of times, earliest and latest, and say `WHERE issued > :earliest AND issued <= :latest`. That way we can be more confident that we have complete coverage of a given range of time. This is especially important, because if a new certificate slips through the cracks for signing, it may potentially be several days before the `staleOCSPResponses` loop catches up with it, since that one is ordered by `issued ASC` as well.